### PR TITLE
fix: fallback to project if addon has no root

### LINKED ISSUE #70: Does not always work with yarn workspaces / monorepos
I'm running into problems with `ember-cli-version-checker` with the following setup: I have a monorepo with an engine, that uses `@ember-decorators/babel-transforms` to enable decorators. The engine uses `ember-cli-babel@7`, which `babel-transforms` checks for using `ember-cli-version-checker`. This works for the engine itself, however it _does not_ work for the engine's dummy app.

Here's why: the dummy app is an instance of `EmberAddon` and has no `root` property. Therefore `basedir` is `undefined`.

https://github.com/ember-cli/ember-cli-version-checker/blob/8443ac7c852aff0be73e1b8687fd912afdc6a7ae/src/npm-dependency-version-checker.js#L14-L16

This means that [`require.sync`](https://github.com/browserify/resolve/blob/3170ce1a3d3a2a14550a4f52b2d201c90d574ce7/lib/sync.js#L39) determines the `basedir` using the physical location of the caller file, `npm-dependency-version-checker.js` in that case:

```js
var basedir = opts.basedir || path.dirname(caller());
```

And this is where it breaks. `ember-cli-version-checker` has been hoisted to the workspace root, consequently `resolve` will start the lookup from there. However, the version of `ember-cli-babel@7` that is used by the engine _has not_ been hoisted. Instead, `ember-cli-babel@6` has been hoisted.

Related to #47.

### PR DESCRIPTION
**Pull request by buschtoens**
Fixes #70.

### CODE REVIEW

File: src/npm-dependency-version-checker.js [hunk 1]
```diff
@@ -12,7 +13,7 @@ class NPMDependencyVersionChecker extends DependencyVersionChecker {
     let jsonPath;
     try {
       jsonPath = resolve.sync(this.name + '/package.json', {
-        basedir: addon.root,
+        basedir: addon.root || getProject(addon).root,
```
Comment (rwjblue) [inline src/npm-dependency-version-checker.js]: I'm not certain this is the correct behavior, do you have a specific reproduction scenario?

Comment (buschtoens) [reply src/npm-dependency-version-checker.js]: I've described a scenario in https://github.com/ember-cli/ember-cli-version-checker/issues/70 😊

**Comment by ef4** [comment]
Thanks @buschtoens, I'm using this patch to workaround the bug for ember-auto-import's test suite (which is now a monorepo, so I can more easily test both babel 6 and babel 7 based apps).

**Comment by alexlafroscia** [comment]
Would it make more sense for the `EmberAddon` object for the dummy app to have a `root` property instead of working around that quirk here?

**Comment by buschtoens** [comment]
If the notion of `root` is applicable to an `EmberAddon`, this would also be a nice fix, I guess. However, that would require affected users to upgrade their host ember-cli version. So maybe we'd wanna do both?

**Comment by buschtoens** [comment]
@rwjblue I've added a test for the `addon` scenario:

> falls back to the project root for instances of `EmberAddon` that do not have a `root` property

I wanted to add a real life workspaces scenario as well that explicitly checks that `resolve.sync` does not fall back to the caller file's physical location, but that is not possible with the current test harness, since the caller file is the test itself.

Let me know, if you want me to change anything.

**Comment by rwjblue** [comment]
> but that is not possible with the current test harness, since the caller file is the test itself

I've _just_ landed some new tests that could show how to do this...

**Comment by buschtoens** [comment]
@rwjblue Thanks for providing such a great basis and pushing for good tests! I've added a real life workspaces scenario.

**Comment by rwjblue** [comment]
Thank you @buschtoens!

**Comment by rwjblue** [comment]
ember-cli-version-checker 3.0.1 published 🎉

**Pull request merged**